cpp-ch/local-engine/Functions/SparkFunctionTrim.cpp

```diff
@@ -106,125 +106,118 @@ namespace
-                res_data[res_offset - 1] = '\0';
-
-                res_offsets[i] = res_offset;
-                prev_offset = offsets[i];
-            }
+            const char * dst;
+            size_t dst_size;
+            trim(src, src_size, dst, dst_size, trim_set);
+            size_t res_offset = row > 0 ? res_offsets[row - 1] : 0;
+            res_data.resize_exact(res_data.size() + dst_size + 1);
+            memcpySmallAllowReadWriteOverflow15(&res_data[res_offset], dst, dst_size);
+            res_offset += dst_size + 1;
+            res_data[res_offset - 1] = '\0';
+            res_offsets[row] = res_offset;
         }
 
-        void executeVector(
-            const ColumnString::Chars & data,
-            const ColumnString::Offsets & offsets,
-            ColumnString::Chars & res_data,
-            ColumnString::Offsets & res_offsets,
-            const ColumnString::Chars & trim_data,
-            const ColumnString::Offsets & trim_offsets) const
+        std::unique_ptr<std::bitset<256>> buildTrimSet(const char* data, const size_t size) const
         {
-            res_data.reserve_exact(data.size());
-
-            size_t rows = offsets.size();
-            res_offsets.resize_exact(rows);
-
-            size_t prev_offset = 0;
-            size_t prev_trim_str_offset = 0;
-            size_t res_offset = 0;
-
-            const UInt8 * start;
-            size_t length;
-
-            for (size_t i = 0; i < rows; ++i)
-            {
-                std::unordered_set<char> trim_set(
-                    &trim_data[prev_trim_str_offset], &trim_data[prev_trim_str_offset] + trim_offsets[i] - prev_trim_str_offset - 1);
-
-                trim(reinterpret_cast<const UInt8 *>(&data[prev_offset]), offsets[i] - prev_offset - 1, start, length, trim_set);
-                res_data.resize_exact(res_data.size() + length + 1);
-                memcpySmallAllowReadWriteOverflow15(&res_data[res_offset], start, length);
-                res_offset += length + 1;
-                res_data[res_offset - 1] = '\0';
-
-                res_offsets[i] = res_offset;
-                prev_offset = offsets[i];
-                prev_trim_str_offset = trim_offsets[i];
-            }
+            auto trim_set = std::make_unique<std::bitset<256>>();
+            for (size_t i = 0; i < size; ++i)
+                trim_set->set((unsigned char)data[i]);
+            return trim_set;
         }
 
-        void
-        trim(const UInt8 * data, size_t size, const UInt8 *& res_data, size_t & res_size, const std::unordered_set<char> & trim_set) const
+        void trim(const char * src, const size_t src_size, const char *& dst, size_t & dst_size, const std::bitset<256> & trim_set) const
         {
-            const char * char_data = reinterpret_cast<const char *>(data);
-            const char * char_end = char_data + size;
+            // If trim column is not constant and contains null value
+            if (trim_set.none())
```

Comment:
trim column doesn't contains null value as long as `useDefaultImplementationForNulls()` returns true. see more details in function `IExecutableFunction::defaultImplementationForNulls`

Reply:
Thanks, removed. defaultImplementationForNulls will use nested column.